### PR TITLE
Adding section on record discriminants

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: content/courses/advanced-ada/parts/data_types/records.rst [hunk 1]
```diff
@@ -941,6 +1374,721 @@ this topic later
...
+       Put_Line ("B'Size = "
+                 & B'Size'Image);
+
+       A := B;  --  STORAGE ERROR!
+    end Show_Storage_Error;
+
+In this case, the assignment :ada:`A := B` raises the :ada:`Storage_Error`
+exception at runtime. Here, the :ref:`storage check <Adv_Ada_Storage_Check>`
+fails because the :ada:`Arr` component of each object has different ranges
+(:ada:`1 .. 10` for object :ada:`A` and :ada:`1 .. 20` for object :ada:`B`).
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/records.rst]: Yes, indeed, that explanation was wrong. I changed it in this commit: https://github.com/AdaCore/learn/pull/1075/commits/40c9db0b7802ce992ae2a5f28c88eb93eaa88497

File: content/courses/advanced-ada/parts/data_types/records.rst [hunk 2]
```diff
@@ -881,12 +881,445 @@ consider using interfaces instead of null records. We'll discuss this topic
...
 
+We introduced the topic of record discriminants in the
+:ref:`Introduction to Ada course <Intro_Ada_Record_Discriminants>`. Also,
+in a previous chapter, we mentioned that record types with unconstrained
+discriminants without defaults are
+:ref:`indefinite types <Adv_Ada_Definite_Indefinite_Subtypes>`.
+
+In this section, we discuss a couple of details about record discriminants that
+we haven't covered yet. Although the discussion will be restricted to
+record discriminants, keep in mind that tasks can also have discriminants.
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/records.rst:892]: Yes, good point!

File: content/courses/advanced-ada/parts/data_types/records.rst [hunk 3]
```diff
@@ -941,6 +1374,721 @@ this topic later
...
+
+
+Derived types with more constrained discriminants
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When deriving types with discriminants, we may use a more constrained type for
+the discriminants of derived type. For example, if the discriminant :ada:`D` of
+the parent type is of :ada:`Integer` type, the corresponding discriminant of
+the derived type may use a constrained subtype such as :ada:`Natural` or
+:ada:`Positive`. For example:
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/records.rst:1861]: Yes, I'll add that to the phrase.

**Comment by gusthoff** [comment]
Thanks a lot for the careful review!

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**